### PR TITLE
fix: Issue #2415

### LINKED ISSUE #2415: Initializing a new array with an initializer function not documented
There are three cases for array initialization, and the reference manual doesn't mention the third (using an initializer function, as in sequence constructions like `seq(5, i => i * 2)`):

```dafny
var a := new int[5];                // Elements initialized with arbitrary values
var b := new int[] [1, 2, 3, 4, 5]; // Elements initialized with the given values
var b := new int[5] (i => i * 2);   // Elements initialized with the given initializer function

```

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #2415

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: docs/DafnyRef/Types.md [hunk 1]
```diff
@@ -2796,8 +2796,13 @@ The initial values of the array elements are arbitrary values of type
...
 a := new T[] [t1, t2, t3, t4];
 ```
+The initial values can also be initialized using a lambda expressions, as follows:
+```
+a := new int[5](i => i*i);
+```
+
 The length of an array is retrieved using the immutable `Length`
-member.  For example, the array allocated above satisfies:
+member.  For example, the array allocated with `a := new int[5];` satisfies:
```
Comment (MikaelMayer) [inline docs/DafnyRef/Types.md:2805]:
```suggestion
member.  For example, the array allocated with `a := new T[n];` satisfies:
```

**Review by MikaelMayer — CHANGES_REQUESTED** [review]
Thanks. I did not know about the lambda initialization for arrays. Very useful.

**Review by robin-aws — CHANGES_REQUESTED** [review]

**Review by MikaelMayer — APPROVED** [review]
you might want to consider @robin-aws 's suggestion as well.

**Comment by davidcok** [comment]
Yes - I added some txt for that also.

**Review by MikaelMayer — APPROVED** [review]

**Review by robin-aws — APPROVED** [review]

**Comment by cpitclaudel** [comment]
I believe this was already documented here:

`````
## 21.16. Array Allocation
````grammar
ArrayAllocation_ =
  "new" [ Type ] "[" [ Expressions ] "]"
  [ "(" Expression(allowLemma: true, allowLambda: true) ")"
  | "[" [ Expressions ] "]"
  ]
````

This allocates a new single or multi-dimensional array as explained in
section [Section 15](#sec-array-types).
The initialization portion is optional. One form is an
explicit list of values, in which case the dimension is optional:
```dafny
var a := new int[5];
var b := new int[5][2,3,5,7,11];
var c := new int[][2,3,5,7,11];
var d := new int[3][4,5,6,7]; // error
```
The comprehension form requires a dimension and uses a function of
type `nat -> T` where `T` is the array element type:
```dafny
var a := new int[5](i => i*i);
```
`````

**Comment by MikaelMayer** [comment]
6 sections later ! We should probably add a link between these sections.

**Pull request merged**